numba_dpex/experimental/launcher.py

```diff
@@ -229,70 +125,85 @@ def get_queue_ref_val(
         for arg_num, argty in enumerate(kernel_argtys):
             if isinstance(argty, DpnpNdArray):
                 llvm_val = kernel_args[arg_num]
-                datamodel = (
-                    self._cpu_codegen_targetctx.data_model_manager.lookup(argty)
-                )
+                datamodel = self.context.data_model_manager.lookup(argty)
                 sycl_queue_attr_pos = datamodel.get_field_position("sycl_queue")
-                ptr_to_queue_ref = self._builder.extract_value(
+                ptr_to_queue_ref = self.builder.extract_value(
                     llvm_val, sycl_queue_attr_pos
                 )
                 break
 
         return ptr_to_queue_ref
 
-    def get_kernel(self, kernel_module, kbref):
+    def get_kernel(self, qref, kernel_module: _KernelModule):
         """Returns the pointer to the sycl::kernel object in a passed in
         sycl::kernel_bundle wrapper object.
         """
-        kernel_name = self._cpu_codegen_targetctx.insert_const_string(
-            self._builder.module, kernel_module.kernel_name
+        # Inserts a global constant byte string in the current LLVM module to
+        # store the passed in SPIR-V binary blob.
+        kernel_bc_byte_str = self.context.insert_const_bytes(
+            self.builder.module,
+            bytes=kernel_module.kernel_bitcode,
         )
-        return sycl.dpctl_kernel_bundle_get_kernel(
-            self._builder, kbref, kernel_name
+
+        # Send it to dpctl, so it create this function on device and returns
+        # reference to it.
+        kbref = self.create_kernel_bundle_from_spirv(
+            queue_ref=qref,
+            kernel_bc=kernel_bc_byte_str,
+            kernel_bc_size_in_bytes=len(kernel_module.kernel_bitcode),
         )
 
+        kernel_name = self.context.insert_const_string(
+            self.builder.module, kernel_module.kernel_name
+        )
+
+        kernel_ref = sycl.dpctl_kernel_bundle_get_kernel(
+            self.builder, kbref, kernel_name
+        )
+
+        sycl.dpctl_kernel_bundle_delete(self.builder, kbref)
+
+        return kernel_ref
+
     def create_llvm_values_for_index_space(
         self,
-        indexer_argty: Union[RangeType, NdRangeType],
-        index_space_arg: llvmir.BaseStructType,
-    ) -> _LLVMIRValuesForIndexSpace:
-        """Returns a list of LLVM IR Values that hold the unboxed extents of a
-        Python Range or NdRange object.
+        ty_indexer_arg: Union[RangeType, NdRangeType],
+        index_arg: llvmir.BaseStructType,
+    ) -> tuple[list, list]:
```

Comment:
Why are you changing from a namedtuple to tuple? Returning a namedtuple makes the code much more clearer to follow as we are basically returning a instance of a class that will give a clear idea of the fields and allow accessing the members in a clean way.

Reply:
My biggest concern was that it does not provide datatype. I've figured out that it is possible to define it in different way:

```python
class class_name(NamedTuple):
    field1: datatype
    field2: datatype
```

That is equivalent to 
```python
class_name = collections.namedtuple('class_name', ['field1', 'field2'])
```
Source: https://www.geeksforgeeks.org/typing-namedtuple-improved-namedtuples/

Reply:
I have nothing against using named tuple if they make sense and does not affect readability/static analysis. In this case it makes sense, but passing named tuple instead of list of arguments seems to be redundant...

Reply:
My biggest pet peeve is that not every code where a function returns a tuple is written in a disciplined way like you originally had.

I have seen way too many places where someone did `result = function_returns_tuple(); a = result[0]; b = result[1]` making the code almost unreadable, especially if the `function_returns_tuple` is deep inside a library.